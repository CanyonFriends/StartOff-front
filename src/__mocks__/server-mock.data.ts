import { ProfileServerType, ProjectServerResponseType, SkillServerType } from '../@types/server';
import generateUUID from '../utils/generateUUID';

export const makeSkillMock = ({
  skill_id = generateUUID(),
  skill_name = 'typescript',
  color = '#eeeeee',
  text_color = '#000000',
}): SkillServerType => {
  return {
    skill_id,
    skill_name,
    color,
    text_color,
  };
};

export const makeProjectResponseMock = ({
  id = Math.floor(Math.random() * 10000),
  title = 'title',
  content = 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s,',
  deploy_url = 'https://shellboylog.com/list',
  github_url = 'https://github.com/qkrdmstlr3/devlog',
  introduce = 'introduce introduce introduce',
  start_date = '1998-01-22',
  end_date = '2021-01-22',
  project_skills = [makeSkillMock({ skill_name: 'typescript' }), makeSkillMock({ skill_name: 'javascript' })],
}): ProjectServerResponseType => {
  return {
    id,
    title,
    content,
    deploy_url,
    github_url,
    introduce,
    start_date,
    end_date,
    project_skills,
  };
};

export const makeProfileMock = ({
  baekjoon_id = 'shellboy',
  blog_url = 'https://shellboylog.com',
  github_url = 'https://github.com/qkrdmstlr3',
  introduce = 'hello my name is shellboy',
  nickname = 'shellboy',
  projects = [makeProjectResponseMock({}), makeProjectResponseMock({})],
  user_skills = [makeSkillMock({ skill_name: 'typescript' }), makeSkillMock({ skill_name: 'javascript' })],
}): ProfileServerType => {
  return {
    baekjoon_id,
    blog_url,
    github_url,
    introduce,
    nickname,
    projects,
    user_skills,
  };
};
