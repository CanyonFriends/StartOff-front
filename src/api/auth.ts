import { AxiosError } from 'axios';
import axios from '../utils/axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  uuid: string;
  email: string;
  nickname: string;
  user_id: string;
}

export interface SignupResponseType {
  user_id: string;
}

export interface LogoutRequestType {
  email: string;
  accessToken: string;
  uuid: string;
}

export interface LogoutResponseType {
  msg: string;
  success: boolean;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType): Promise<ErrorType | SignupResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response.data as SignupResponseType;
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const logoutAPI = async ({
  email,
  accessToken,
  uuid,
}: LogoutRequestType): Promise<ErrorType | LogoutResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/logout',
      data: {
        email,
        uuid,
        access_token: accessToken,
      },
    });
    return response.data as LogoutResponseType;
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};
