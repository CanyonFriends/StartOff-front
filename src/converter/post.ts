import { BoardServerType, SummarizedPostServerType, CreatePostServerType, PostServerType } from '../@types/server';
import { BoardClientType, SummarizedPostClientType, CreatePostClientType, PostClientType } from '../@types/client';
import { skillServerType2ClientType } from './skill';

export const summarizedPostServerType2ClientType = (
  summarizedPost: SummarizedPostServerType,
): SummarizedPostClientType => {
  return {
    postId: String(summarizedPost.post_id),
    title: summarizedPost.title,
    currentPeople: summarizedPost.current_people,
    maxPeople: summarizedPost.max_people,
    createAt: new Date(summarizedPost.created_at),
    nickname: summarizedPost.nickname,
    postSkills: summarizedPost.post_skills.map((skill) => skillServerType2ClientType(skill)),
  };
};

export const boardServerType2ClientType = (board: BoardServerType): BoardClientType => {
  return {
    content: board.content.map((post) => summarizedPostServerType2ClientType(post)),
    totalElements: board.totalElements,
    totalPages: board.totalPages,
  };
};

export const createPostClientType2ServerType = (post: CreatePostClientType): CreatePostServerType => {
  return {
    category: post.category.toUpperCase(),
    content: post.content,
    current_people: post.currentPeople,
    max_people: post.maxPeople,
    post_skills: post.postSkills.map((skill) => skill.skillName),
    title: post.title,
    user_id: Number(post.userId),
  };
};

export const postServerType2ClientType = (post: PostServerType): PostClientType => {
  return {
    postId: String(post.post_id),
    category: post.category,
    title: post.title,
    content: post.content,
    createdAt: new Date(post.created_at),
    currentPeople: post.current_people,
    maxPeople: post.max_people,
    nickname: post.nickname,
    postSkills: post.post_skills.map((skill) => skillServerType2ClientType(skill)),
    userId: String(post.user_id),
  };
};
