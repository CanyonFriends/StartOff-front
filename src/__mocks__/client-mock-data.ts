import { ErrorType } from '../api/error';
import { SigninResponseType } from '../api/auth';
import {
  SkillClientType,
  ProjectClientType,
  ProfileClientType,
  SummarizedPostClientType,
  BoardClientType,
  CreatePostClientType,
  PostClientType,
} from '../@types/client';
import generateUUID from '../utils/generateUUID';

// api
export const makeSigninSuccessInfoMock = ({
  access_token = 'access_token',
  uuid = 'uuid',
  email = 'email@email.com',
  nickname = 'nickname',
  user_id = '1',
}): SigninResponseType => {
  return {
    access_token,
    uuid,
    email,
    nickname,
    user_id,
  };
};

// error
export const makeErrorMock = ({ error_msg = '에러 발생' }): ErrorType => {
  return { error_msg };
};

// data
export const makeSkillMock = ({
  skillId = generateUUID(),
  skillName = 'c#',
  color = '#eeeeee',
  textColor = '#000000',
}): SkillClientType => {
  return {
    skillId,
    skillName,
    color,
    textColor,
  };
};

export const makeProjectMock = ({
  id = Math.floor(Math.random() * 10000),
  title = 'project',
  deployUrl = 'https://shellboylog.com/list',
  githubUrl = 'https://github.com/qkrdmstlr3/devlog',
  introduce = 'introduce introduce introduce',
  startDate = new Date('1998-01-22'),
  endDate = new Date('2021-01-22'),
  content = 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s,',
  projectSklls = [makeSkillMock({ skillName: 'typescript' }), makeSkillMock({ skillName: 'c#' })],
}): ProjectClientType => {
  return {
    id,
    title,
    deployUrl,
    githubUrl,
    introduce,
    startDate,
    endDate,
    content,
    projectSklls,
  };
};

export const makeProfileMock = ({
  baekjoonId = 'shellboy',
  blogUrl = 'https://shellboylog.com',
  githubUrl = 'https://github.com/qkrdmstlr3',
  introduce = 'hello my name is shellboy',
  nickname = 'shellboy',
  projects = [makeProjectMock({})],
  userSkills = [makeSkillMock({ skillName: 'typescript' }), makeSkillMock({ skillName: 'c#' })],
}): ProfileClientType => {
  return {
    baekjoonId,
    blogUrl,
    githubUrl,
    introduce,
    nickname,
    projects,
    userSkills,
  };
};

export const makeMockSummarizedPost = ({
  postId = generateUUID(),
  title = 'title',
  currentPeople = 5,
  maxPeople = 10,
  createAt = new Date(),
  nickname = 'user',
  postSkills = [makeSkillMock({ skillName: 'typescript' }), makeSkillMock({ skillName: 'c#' })],
}): SummarizedPostClientType => {
  return {
    postId,
    title,
    currentPeople,
    maxPeople,
    createAt,
    postSkills,
    nickname,
  };
};

export const makeMockBoard = ({
  content = [makeMockSummarizedPost({})],
  totalElements = 1,
  totalPages = 1,
}): BoardClientType => {
  return {
    content,
    totalElements,
    totalPages,
  };
};

export const makeCreatePost = ({
  category = 'category',
  content = 'content',
  currentPeople = 5,
  maxPeople = 10,
  postSkills = [makeSkillMock({})],
  title = 'title',
  userId = '1',
}): CreatePostClientType => {
  return {
    category,
    content,
    currentPeople,
    maxPeople,
    postSkills,
    title,
    userId,
  };
};

export const makePostMock = ({
  postId = generateUUID(),
  category = 'category',
  title = 'title',
  content = 'content',
  createdAt = new Date('1998-01-22'),
  currentPeople = 5,
  maxPeople = 10,
  nickname = 'shellboy',
  postSkills = [makeSkillMock({})],
}): PostClientType => {
  return {
    postId,
    category,
    title,
    content,
    createdAt,
    currentPeople,
    maxPeople,
    nickname,
    postSkills,
  };
};

export const mockString =
  "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";
