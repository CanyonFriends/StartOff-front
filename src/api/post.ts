import axios from '../utils/axios';
import { ErrorType } from './error';
import { BoardServerType } from '../@types/server';
import { boardServerType2ClientType, createPostClientType2ServerType } from '../converter/post';
import { CreatePostClientType } from '../@types/client';

export const getCategoriesAPI = async () => {
  try {
    const response = await axios({
      method: 'GET',
      url: '/v1/categories',
    });

    return response.data as string[];
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const getPostsAPI = async (page: number, size: number, category: string) => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/posts?page=${page}&size=${size}&category=${category}`,
    });

    return boardServerType2ClientType(response.data as BoardServerType);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const createPostAPI = async (post: CreatePostClientType) => {
  try {
    await axios({
      method: 'POST',
      url: '/v1/posts',
      data: {
        ...createPostClientType2ServerType(post),
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
