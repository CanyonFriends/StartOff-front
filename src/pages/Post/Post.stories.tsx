import React from 'react';
import { makePostMock } from '../../__mocks__/client-mock-data';
import PostTemplate from './template';

export default {
  title: 'Template/Post',
  component: PostTemplate,
};

const postMock = makePostMock({});

export const editable = () => <PostTemplate post={postMock} editableAuthority />;
export const unEditable = () => <PostTemplate post={postMock} editableAuthority={false} />;
