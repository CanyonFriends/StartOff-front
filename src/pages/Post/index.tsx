import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import PostTemplate from './template/index';
import { PostClientType } from '../../@types/client';
import { getPostAPI } from '../../api/post';
import { isFailed } from '../../api/error';
import { RootState } from '../../redux/store';
import { UserState } from '../../redux/user/types';

interface ParamProps {
  postId: string;
}

function Post() {
  const { postId } = useParams<ParamProps>();
  const [post, setPost] = useState<PostClientType>();
  const userState = useSelector<RootState>((state) => state.user) as UserState;

  useEffect(() => {
    getPost();
  }, []);

  const getPost = async () => {
    const response = await getPostAPI(postId);
    if (isFailed<PostClientType>(response)) {
      return;
    }
    setPost(response);
  };

  return post ? <PostTemplate post={post} editableAuthority={userState.userId === post.userId} /> : <></>;
}

export default Post;
