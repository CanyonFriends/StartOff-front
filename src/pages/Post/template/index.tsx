import React from 'react';
import { useHistory } from 'react-router-dom';
import { PostClientType } from '../../../@types/client';
import CommonHeader from '../../../components/Layout/CommonHeader';
import { PostTemplate } from '../../../components/UI/organism';
import * as Style from './styled';
import { deletePostAPI } from '../../../api/post';
import { isFailed } from '../../../api/error';
import { buildBoardPath, buildModifyPath } from '../../../Routes';

interface PostTemplateProps {
  editableAuthority: boolean;
  post: PostClientType;
}

function PostPostTemplate({ editableAuthority, post }: PostTemplateProps) {
  const history = useHistory();
  const handleModifyPost = async () => {
    history.push(buildModifyPath(post.postId));
  };

  const handleDeletePost = async () => {
    // TODO: confirm으로 확인 한번 받기
    const response = await deletePostAPI(post.postId);
    if (isFailed<boolean>(response)) {
      return;
    }
    history.push(buildBoardPath(post.category, 0));
  };

  return (
    <>
      <CommonHeader />
      <Style.Container>
        <Style.PostWrapper>
          <PostTemplate
            editableAuthority={editableAuthority}
            title={post.title}
            content={post.content}
            currentPeople={post.currentPeople}
            maxPeople={post.maxPeople}
            postSkills={post.postSkills}
            nickname={post.nickname}
            handleModifyPost={handleModifyPost}
            handleDeletePost={handleDeletePost}
          />
        </Style.PostWrapper>
      </Style.Container>
    </>
  );
}

export default PostPostTemplate;
