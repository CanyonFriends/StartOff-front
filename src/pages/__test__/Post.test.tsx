/**
 * @jest-environment jsdom
 */
import React from 'react';
import { waitFor } from '@testing-library/react';
import { render } from '../../test-utils';
import Post from '../Post';
import { getCategoriesAPI, getPostAPI } from '../../api/post';
import { makePostMock } from '../../__mocks__/client-mock-data';

jest.mock('../../api/post');

const getCategoriesMockAPI = getCategoriesAPI as jest.MockedFunction<typeof getCategoriesAPI>;
const getPostMockAPI = getPostAPI as jest.MockedFunction<typeof getPostAPI>;

beforeEach(() => {
  jest.resetAllMocks();
  getCategoriesMockAPI.mockReturnValue(new Promise((res) => res([])));
  getPostMockAPI.mockReturnValue(new Promise((res) => res(makePostMock({ nickname: undefined }))));
});

describe('<Post> 페이지', () => {
  it('snapshot 체크', async () => {
    const component = render(<Post />);
    await waitFor(() => expect(component.container).toMatchSnapshot());
  });
});
