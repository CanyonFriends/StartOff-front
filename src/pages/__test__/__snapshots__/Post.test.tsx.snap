// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post> 페이지 snapshot 체크 1`] = `
<div>
  <div
    id="modal"
  />
</div>
`;
